Code/src/megadesk.cpp

```diff
@@ -657,72 +726,81 @@ void saveMemory(uint8_t memorySlot, uint16_t value)
   if (memorySlot < 2 || value < 5 || value > 32700)
     return;
 
-  //beep(memorySlot, NOTE_HIGH);
+  // save confirmation tone
+  beep(NOTE_HIGH);
 
-  eeprom_put16(memorySlot, value);
+  eepromPut16(memorySlot, value);
 }
 
 uint16_t loadMemory(uint8_t memorySlot)
 {
   if (memorySlot < 2)
     return currentHeight;
 
-  beep(memorySlot, NOTE_LOW);
-
-  uint16_t memHeight = eeprom_get16(memorySlot);
+  uint16_t memHeight = eepromGet16(memorySlot);
 
   if (memHeight == 0)
   {
     // empty
     delay(LONG_PAUSE);
     // sad trombone
-    beep(1, NOTE_DSHARP7);
-    beep(1, NOTE_D7);
-    beep(1, NOTE_CSHARP7);
-    beep(1, NOTE_C7);
+    beep(NOTE_DSHARP7);
+    beep(NOTE_D7);
+    beep(NOTE_CSHARP7);
+    beep(NOTE_C7);
   }
 
   return memHeight;
 }
 
-void delay_until(unsigned long microSeconds)
+
+// accurate delay from the last time delayUntil() returned.
+// for precise periodic timing
+void delayUntil(unsigned long microSeconds)
 {
-  unsigned long end = t + (1000 * microSeconds);
-  unsigned long d = end - micros();
+  unsigned long target = refTime + (1000 * microSeconds);
+  unsigned long micro_delay = target - micros();
 
-  // crazy long delay; probably negative wrap-around
-  // just return
-  if (d > 1000000)
+  if (micro_delay > 1000000)
   {
-    t = micros();
+    // crazy long delay - target time is in the past!
+    // reset refTime and return
+    refTime = micros();
     return;
   }
 
-  if (d > 15000)
+  if (micro_delay > 15000)
   {
-    unsigned long d2 = (d - 15000) / 1000;
-    delay(d2);
-    d = end - micros();
+    // need delayMicroseconds() and delay()
+    unsigned long millidelay = (micro_delay - 15000) / 1000;
+    delay(millidelay);
+    // recalculate microsec delay
+    micro_delay = target - micros();
   }
-  delayMicroseconds(d);
-  t = end;
+  delayMicroseconds(micro_delay);
+  refTime = target;
 }
 
-// simple tone generation - leaner than tone()
-// sound will break up if servicing ISRs/interrupts.
+// simple, limited tone generation - leaner than tone()
+// but sound will break up if servicing interrupts.
+// freq is in Hz. duration is in ms. (max 524ms)
 void playTone(uint16_t freq, uint16_t duration) {
-  uint16_t halfperiod = 1000000L / freq;
-  for (long i = 0; i < duration * 1000L; i += halfperiod * 2) {
+  uint16_t halfperiod = 1000000L / freq; // in us.
+  // mostly equivalent to:
+  // for (long i = 0; i < duration * 1000L; i += halfperiod * 2) {
+  for ( duration = (125 * duration) / (halfperiod / 4); duration > 0; duration -= 1 ) {
```

Comment:
There was 100byte saving of image size by avoiding long int manipulation every loop, the tuning improved too.

This freaky order of arithmetic is to ensure we can specify up to 500ms duration without overflowing the uint16 before the division. Other order of operations overflowed or exhibited inaccurate/frequency-dependent duration!

The duration isn't completely accurate but this is a desk not a synthesizer, so no-one should notice.